dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -70,4 +116,59 @@ public void run(Function function) {
             counter++;
         }
     }
+
+    private void guaranteeSingleEntry(Function function) {
+        int labelCounter = 0;
+        for (Label label : function.getEvents(Label.class)) {
+
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            if (backJumps.isEmpty()) {
+                continue;
+            }
+
+            int counter = 0;
+
+            final Label loopBegin = label;
+            final CondJump uniqueBackJump = backJumps.get(0);
+
+            final Register sourceReg = function.newRegister(String.format("__jumpedTo_%s#%s_From", label, labelCounter), types.getArchType());
+            final Local initJumpedFromOutside = EventFactory.newLocal(sourceReg, expressions.makeZero(types.getArchType()));
+            function.getEntry().insertAfter(initJumpedFromOutside);
```

Comment:
This is insufficient. You need to reset right before the loop is entered (right before the label) not only at the beginning of the function.
Consider the case of two loops `A` and `B` both of which have an outside jump.
The following can happen:
(1) A is entered from the outside, setting the entry reg to `1`.
(2) A terminates in the same iteration it was entered: the backjump is not taken and the register is not reset.
(3) B is entered through the regular way, but the entry reg still contains `1`. B accidentally thinks it was entered from the outside.

Reply:
If `A` and `B` are different loops, we will use different registers so this should not be an issue.

I decided to place the initialization at the beginning of the function because we might have code like this
```
goto A
A:
```
Putting the initialization just before the label would not work and I wanted to avoid having to consider this special case.